java/arcs/core/allocator/Allocator.kt

```diff
@@ -138,18 +143,26 @@ class Allocator private constructor(
         }
     }
 
-    /**
-     * Reads associated [PlanPartition]s with an [ArcId] .
-     */
-    private suspend fun readPartitionMap(arcId: ArcId): List<Plan.Partition>? {
-        val entities = collection.fetchAll().filter { fieldAsString(it, "arc") == arcId.toString() }
-        return entities.map {
-            Plan.Partition(
-                arcId.toString(),
-                fieldAsString(it, "host"),
-                fieldAsStrings(it, "particles").map { Plan.Particle(it, "", mapOf()) }
-            )
-        }
+    /** Looks up [RawEntity]s representing [PlanPartition]s for a given [ArcId] */
+    private suspend fun entitiesForArc(arcId: ArcId): List<RawEntity> =
+        collection.fetchAll().filter { fieldAsString(it, "arc") == arcId.toString() }
+
+    /** Converts a [RawEntity] to a [Plan.Partition] */
+    private fun entityToPartition(entity: RawEntity): Plan.Partition =
+        Plan.Partition(
+            fieldAsString(entity, "arc"),
+            fieldAsString(entity, "host"),
+            fieldAsStrings(entity, "particles").map { Plan.Particle(it, "", mapOf()) }
+        )
+
+    /** Reads associated [PlanPartition]s with an [ArcId]. */
+    private suspend fun readPartitions(arcId: ArcId): List<Plan.Partition> =
+        entitiesForArc(arcId).map { entityToPartition(it) }
+
+    private suspend fun readAndClearPartitions(arcId: ArcId): List<Plan.Partition> {
+        val entities = entitiesForArc(arcId)
+        entities.forEach { collection.remove(it) }
+        return entities.map { entityToPartition(it) }
```

Comment:
You could chain all of these:

```
= entitiesForArc(arcId)
    .forEach { collection.remove(it) }
    .map { entityToPartition(it) }
```

Reply:
oooh that's quite cute.

Reply:
oh, but it doesn't work :( forEach returns Unit.